ggml-cuda.cu

```diff
@@ -2405,11 +2405,100 @@ GGML_CALL static void ggml_backend_cuda_synchronize(ggml_backend_t backend) {
     GGML_UNUSED(backend);
 }
 
+#if (CUDART_VERSION >= 12000)
+#define USE_CUDA_GRAPH
+#endif
+
+#ifdef USE_CUDA_GRAPH
+#define MAX_NODES_IN_CUDA_GRAPH 10000
+struct ggml_cudaGraph {
+    int count=0;
+    cudaGraph_t graph = nullptr;
+    cudaGraphExec_t instance = nullptr;
+    size_t numNodes = 0;
+    int softmax_ne0 = 0;
+    cudaGraphNode_t nodes[MAX_NODES_IN_CUDA_GRAPH];
+    cudaKernelNodeParams params[MAX_NODES_IN_CUDA_GRAPH];
+    bool disableDueToGpuArch=false;
+};
+#endif
+
+const bool disableCudaGraphs = (getenv("LLAMACPP_DISABLE_CUDA_GRAPHS") != nullptr);
+
 GGML_CALL static enum ggml_status ggml_backend_cuda_graph_compute(ggml_backend_t backend, ggml_cgraph * cgraph) {
     ggml_backend_cuda_context * cuda_ctx = (ggml_backend_cuda_context *)backend->context;
 
     ggml_cuda_set_device(cuda_ctx->device);
 
+#ifdef USE_CUDA_GRAPH
+    // Objects required for CUDA Graph
+    static ggml_cudaGraph cudaGraph;
+    bool useCudaGraph = (cudaGraph.count>=7); //avoid CUDA graphs on first few steps due to incompatible initialisations.
+    char** updatedKernelArg[MAX_NODES_IN_CUDA_GRAPH];
+    bool cudaGraphUpdateRequired = false;
+    // pointer to CUDA cpy kernel, which is required to identify
+    // kernel parameters which need updated in the graph for each token
+    void* ggmlCudaCpyFn = nullptr;
+
+    if(cudaGraph.count==0){        
+        cudaDeviceProp prop;
+        int device;
+        CUDA_CHECK(cudaGetDevice(&device));
```

Comment:
Use `ggml_cuda_get_device` and `ggml_cuda_set_device`. For getting the device this doesn't actually matter but when setting the device llama.cpp has a workaround for Windows to make the performance less bad compared to Linux.

Also I think it would be better to just get the device from `cuda_ctx`. Is that correct @slaren ?

Reply:
Done (using cuda_ctx)